compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -543,7 +545,32 @@ impl UnsafeOpKind {
         tcx: TyCtxt<'_>,
         hir_id: hir::HirId,
         span: Span,
+        suggest_unsafe_block: bool,
     ) {
+        let note_non_inherited = tcx.hir().parent_iter(hir_id).find(|(id, _)| {
+            if let Some(sig) = tcx.hir().fn_sig_by_hir_id(*id)
+                && sig.header.is_unsafe()
+            {
+                true
+            } else {
+                false
+            }
+        });
+        let unsafe_not_inherited_note = if let Some((id, _)) = note_non_inherited {
+            suggest_unsafe_block.then(|| {
+                let span = tcx.hir().span(id);
+                let signature_span = tcx.sess.source_map().guess_head_span(span);
+                let body = tcx.hir().body_owned_by(hir_id.owner.def_id);
+                let body_span = tcx.hir().body(body).value.span;
+                UnsafeNotInheritedLintNote {
+                    signature_span,
+                    body_start: tcx.sess.source_map().start_point(body_span).shrink_to_hi(),
+                    body_end: tcx.sess.source_map().end_point(body_span).shrink_to_lo(),
```

Comment:
Having to keep in sync the logic between here and the `impl AddToDiagnostic` is not great, but I don't really have a better suggestion either. Only keep `body_span` in the struct, output `{ unsafe ` at `body_span.shrink_to_lo()`,  and ` }` at `body_span.shrink_to_hi()`?